cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -1005,4 +1013,57 @@ private ClassLoader setContextCombinedClassLoader(BasicMapReduceContext context)
     return ClassLoaders.setContextClassLoader(new CombineClassLoader(
       null, ImmutableList.of(context.getProgram().getClassLoader(), getClass().getClassLoader())));
   }
+
+  /**
+   * Localizes resources requested by users in the MapReduce Program's beforeSubmit phase.
+   * In Local mode, also copies resources to a temporary directory.
+   *
+   * @param job the {@link Job} for this MapReduce program
+   * @param targetDir in local mode, a temporary directory to copy the resources to
+   * @return a {@link Map} of resource name to the {@link File} handle for the local file.
+   */
+  private Map<String, File> localizeUserResources(Job job, File targetDir) throws IOException {
+    Map<String, File> localizedResources = new HashMap<>();
+    Map<String, LocalizeResource> resourcesToLocalize = context.getResourcesToLocalize();
+    for (final Map.Entry<String, LocalizeResource> entry : resourcesToLocalize.entrySet()) {
+      final File localizedFile;
+      LocalizeResource resourceToLocalize = entry.getValue();
+      URI uri = resourceToLocalize.getURI();
+      String name = entry.getKey();
+      Configuration mapredConf = job.getConfiguration();
+      if (MapReduceTaskContextProvider.isLocal(mapredConf)) {
+        // in local mode, also add localize resources in a temporary directory
+        File tempFile = new File(targetDir, name);
+        localizedFile = new File(uri.getPath());
+        LOG.debug("Copy file from {} to {}", uri, tempFile);
+        Files.copy(new InputSupplier<InputStream>() {
+          @Override
+          public InputStream getInput() throws IOException {
+            return new FileInputStream(localizedFile);
+          }
+        }, tempFile);
+      } else {
+        // in distributed mode, use the MapReduce Job object to localize resources
+        URI actualURI;
+        try {
+          actualURI = new URI(uri.getScheme(), uri.getAuthority(), uri.getPath(), uri.getQuery(), name);
+        } catch (URISyntaxException e) {
+          // Most of the URI is constructed from the passed URI. So ideally, this should not happen.
+          // If it does though, there is nothing that clients can do to recover, so not propagating a checked exception.
+          throw Throwables.propagate(e);
+        }
+        if (resourceToLocalize.isArchive()) {
+          job.addCacheArchive(actualURI);
+        } else {
+          job.addCacheFile(actualURI);
+        }
+        localizedFile = new File(name);
+      }
+      localizedResources.put(name, localizedFile);
+      // also serialize the local resource names, so they can be recreated later.
+      mapredConf.set(Constants.Localization.LOCAL_RESOURCES, Joiner.on(",").join(resourcesToLocalize
```

Comment:
Why needed? I think you already serialize the map through the `MapReduceContextConfig` class?